### PR TITLE
feat(app-platform): Refresh Token exchange

### PR DESCRIPTION
**Pull request by mnoble**
Augments the Installation authorization endpoint to allow refreshing of tokens. If the token isn't already expired, this process expires it and returns a new one. If it's already expired, it just returns a new one.

### CODE REVIEW

File: src/sentry/mediators/sentry_app_installations/authorizer.py [hunk 1]
```diff
@@ -13,31 +14,53 @@
...
+    def create_api_token(self):
         return ApiToken.objects.create(
             user=self.user,
             application=self.application,
             scope_list=self.sentry_app.scope_list,
-            expires_at=(datetime.now() + timedelta(hours=TOKEN_LIFE_IN_HOURS)),
+            expires_at=(datetime.utcnow() + timedelta(hours=TOKEN_LIFE_IN_HOURS)),
         )
 
+    def _validate_code_or_refresh_token(self):
```
Comment (MeredithAnya) [inline src/sentry/mediators/sentry_app_installations/authorizer.py]: would it make sense to combine `_validate_code_or_refresh_token` and `_validate_grant_type` and do something like 
```python
is_refresh_token = self.refresh_token and self.grant_type == 'refresh_token' 
is_authorization_code = self.code and self.grant_type == 'authorization_code' 

if not is_refresh_token or is_authorization_code:
    raise APIUnauthorized
```

cause what happens if someone accidentally sends `code` and `refresh_token`

File: src/sentry/mediators/sentry_app_installations/authorizer.py [hunk 2]
```diff
@@ -95,3 +132,10 @@ def grant(self):
                 .get(code=self.code)
         except ApiGrant.DoesNotExist:
             raise APIUnauthorized
+
+    @property
+    def token(self):
+        try:
+            return ApiToken.objects.get(refresh_token=self.refresh_token)
```
[the remark this answers is not in the document]
Comment (mnoble) [reply src/sentry/mediators/sentry_app_installations/authorizer.py]: No, but they're UUIDs, so the chances are slim. Couldn't hurt to add an actual constraint, though.

**Review by markstory — APPROVED** [review]
Looks good to me. I like the separation of the code and refresh flows. It makes both easier to understand.

File: src/sentry/mediators/sentry_app_installations/authorizer.py [hunk 3]
```diff
@@ -59,6 +82,20 @@ def _validate_grant(self):
         if self.grant.is_expired():
             raise APIUnauthorized
 
+    def _validate_refresh_token(self):
+        if (
+            self.token.application != self.application or
+            self.is_token_expired()
+        ):
+            raise APIUnauthorized
```
[the remark this answers is not in the document]
Comment (mnoble) [reply src/sentry/mediators/sentry_app_installations/authorizer.py]: I moved the expired grant case out and added a corresponding error below. The other two conditions above, I think, are still okay as generic "you're doing something you shouldn't be".

[the remark this answers is not in the document]
Comment (markstory) [reply src/sentry/mediators/sentry_app_installations/authorizer.py]: :+1:

**Pull request merged**